frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -1651,9 +1658,27 @@ impl<T: Config> ElectionProvider for Pallet<T> {
 				Ok(supports)
 			},
 			Err(why) => {
-				log!(error, "Entering emergency mode: {:?}", why);
-				<CurrentPhase<T>>::put(Phase::Emergency);
-				Err(why)
+				// if `elect()` was called during `Phase::Off` or too early since the beginning of
+				// the signed phase, do not enter in emergency mode since there was no enough time
+				// to reach a solution.
+				match CurrentPhase::<T>::get() {
```

Comment:
I think this is a good solution. 

Just food for thought, should we not call `do_elect` (line 1653) at all if either (1) phase is off, or (2) phase::signed started not too long ago? That would mean even fallback is also not called and we are making a strong decision that epm is not ready to have an election right now.

I like the current approach too. `We try our best to do an election, but if it does not work out, we are not going to panic just yet`.

Reply:
Yes that's a good option too. I think something in the middle would probably be a good way fwd:
 
Try the election and then:
- If it fails and `Phase::Off` || `Phase::Signed` for few blocks only, fail election but not set emergency mode and do not try the fallback election just yet.
- If it fails and `Phase::Signed` has been running for enough time, try the fallback election. If it fails, enters in emergency mode.

I think the advantage of this is that EPM doesn't start the fallback election unless the normal election had enough time to run. The fallback election may be expensive, so we probably want to postpone it as much as possible within reason (similar to jumping to emergency phase).

wdyt? cc @kianenigma

Reply:
The [latest commit](https://github.com/paritytech/substrate/pull/13040/commits/7ffcd18e8e99e72c77a095693c9e28a3c9f8d45c) implements the logic mentioned above.

Reply:
I am not too sure about this. We are basically throttling on `fallback` but the point of fallback is to be executed when `for any reason`, normal `elect` does not succeed. 

Not going to emergency is reasonable I feel (your previous changes) ✅. 
Not doing election at all if `too_early` (my suggestion) is also probably wrong as we will be undermining the role of `fallback`. 

Kian's voice is now echoing in my head that we just need a good fallback. :D

Reply:
I agree with you that we do need a sound fallback strategy (perhaps we should start thinking about it too). My reasoning behind this flow is that if we don't expect the normal election to go through (and thus not trigger the emergency phase), we should also not expect the fallback to run (so we spare the resources required to run the on-chain fallback).

So we run the fallback and eventually go to emergency _only_ if we're sure the election had enough time to complete but failed for some reason.

Reply:
There are three steps (in the worst case)
- Normal Election
- Fallback
- Emergency

What I was arguing, `Fallback` is a fallback of `Normal Election` and so (Normal Election + fallback) should be seen as one component. That is, whenever `Normal Election` is called and fails, we should call `Fallback` (because its the fallback :D).

Based on this, my suggestion is either not call `Normal election` at all if EPM think its called abruptly and its not prepared, or call the whole unit (`Normal Election` + `Fallback`) and then throttle (your older proposal).